plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs

```diff
@@ -0,0 +1,151 @@
+
+-- See Note [Cost model parameters]
+-- | Extract the model parameters from a model.
+extractParams :: ToJSON a => a -> Maybe CostModelParams
+extractParams cm = case toJSON cm of
+    Object o ->
+        let
+            flattened = flattenObject "-" o
+            toScaledInteger :: S.Scientific -> Integer
+            toScaledInteger n = floor (n*1000*1000)
+            scaledNumbers = HM.mapMaybe (\case { Number n -> Just $ toScaledInteger n; _ -> Nothing }) flattened
+            mapified = Map.fromList $ HM.toList scaledNumbers
+        in Just mapified
+    _ -> Nothing
+
+
+-- See Note [Cost model parameters]
+-- | Update a model by overwriting the parameters with the given ones.
+applyParams :: (FromJSON a, ToJSON a) => a -> CostModelParams -> Maybe a
+applyParams cm params = case toJSON cm of
+    Object o ->
+        let
+            hashmapified = HM.fromList $ Map.toList params
+            scaledNumbers = fmap (\n -> Number $ fromIntegral n / (1000*1000)) hashmapified
+            flattened = flattenObject "-" o
+            -- this is where the overwriting happens, this is left-biased
+            merged = HM.union scaledNumbers flattened
+            unflattened = unflattenObject "-" merged
+        in case fromJSON (Object unflattened) of
+            Success a -> Just a
+            Error _   -> Nothing
+    _ -> Nothing
+
+-- | Parameters for a machine step model and a builtin evaluation model bundled together.
+data SplitCostModelParams =
+    SplitCostModelParams {
+      machineParams :: CostModelParams
+    , builtinParams :: CostModelParams
+    }
+
+-- | Split a CostModelParams object into two subobjects according to some prefix:
+-- see item 5 of Note [Cost model parameters].
+splitParams :: Text.Text -> CostModelParams -> SplitCostModelParams
+splitParams prefix params =
+    let machineparams = Map.filterWithKey (\k _ ->       Text.isPrefixOf prefix k) params
+        builtinparams = Map.filterWithKey (\k _ -> not $ Text.isPrefixOf prefix k) params
+    in SplitCostModelParams machineparams builtinparams
+
+-- | Given a CostModel, produce a single map containing the parameters from both components
+extractCostModelParams :: ToJSON machinecosts => CostModel machinecosts -> Maybe CostModelParams
+extractCostModelParams model =
+    case ( extractParams (machineCostModel model)
+         , extractParams (builtinCostModel model) )
+    of (Just machineparams, Just builtinparams) -> Just $ Map.union machineparams builtinparams
+       _                                        -> Nothing
+
+-- | Given a set of cost model parameters, split it into two parts according to some
+-- prefix and use those parts to update the components of a cost model.
+applySplitCostModelParams
+    :: (FromJSON evaluatorcosts, ToJSON evaluatorcosts)
+    => Text.Text
+    -> CostModel evaluatorcosts
+    -> CostModelParams
+    -> Maybe (CostModel evaluatorcosts)
+applySplitCostModelParams prefix model params =
```

Comment:
There's quite a lot of messing about with different structures here (and in this file in general).  It might be worth trying to make things more efficient by doing things all in one go.